File: memcontainers/shcore/src/expand.rs
//! Word expansion — the POSIX expansion pipeline.
//!
//! Applied in order: tilde → parameter / command / arithmetic expansion → field
//! splitting (on IFS) → pathname (glob) expansion → quote removal (quote removal
//! already happened during lexing, recorded in each fragment's quoting flags).
//!
//! The driver targets the [`ExpandCtx`] trait so it stays OS-agnostic: the executor
//! supplies variable lookup, command substitution, arithmetic, and directory listing.
//! The subtle invariant threaded through here is *quoting*: a quoted expansion is
//! pushed un-split with glob inactive, an unquoted one is split on IFS with glob
//! active — so `"$x"` is one field that never globs while `$x` splits and may glob.
//!
//! Errors (a `${x:?msg}` violation) propagate as [`ExpandError`] via `Result`, so a
//! caller can never forget to check them — there is no stateful error flag. Integrated
//! behaviour (real `$()`, real globbing) is validated by the e2e suite.

use alloc::string::{String, ToString};
use alloc::vec::Vec;

use crate::glob::expand_glob_masked;
use crate::word::{ParamOp, Word, WordPart};

/// An expansion failure carrying the message to report (e.g. from `${x:?msg}`).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ExpandError(pub String);

type R<T> = Result<T, ExpandError>;

/// Everything expansion needs from the outside world.
pub trait ExpandCtx {
    /// Ordinary shell variable (None if unset).
    fn get(&mut self, name: &str) -> Option<String>;
    /// Set a shell variable (for `${x:=word}`).
    fn set(&mut self, name: &str, val: &str);
    /// Scalar special parameters: `?` `$` `!` `#` `-` and `0`..`9`.
    fn special(&mut self, name: &str) -> Option<String>;
    /// Positional parameters `$1 $2 …` (for `$@` / `$*`).
    fn positionals(&mut self) -> Vec<String>;
    /// Run a command string and return its stdout (trailing newlines stripped).
    fn command_subst(&mut self, raw: &str) -> String;
    /// Evaluate an arithmetic expression.
    fn arith(&mut self, expr: &str) -> i64;
    /// List a directory (for globbing); None if unreadable.
    fn list_dir(&mut self, path: &str) -> Option<Vec<String>>;
    /// Current working directory (for relative globs).
    fn cwd(&mut self) -> String;
    /// Current IFS (defaults to space/tab/newline when unset).
    fn ifs(&mut self) -> String;
    /// `$HOME` for tilde expansion.
    fn home(&mut self) -> Option<String>;
}

fn lookup(ctx: &mut dyn ExpandCtx, name: &str) -> Option<String> {
    if let Some(v) = ctx.special(name) {
        return Some(v);
    }
    ctx.get(name)
}

// ---------- field builder (handles $@ field-injection) ----------

struct FieldBuilder {
    fields: Vec<(Vec<char>, Vec<bool>)>, // (chars, glob-active mask)
    cur_c: Vec<char>,
    cur_a: Vec<bool>,
    started: bool,
}

impl FieldBuilder {
    fn new() -> Self {
        FieldBuilder {
            fields: Vec::new(),
            cur_c: Vec::new(),
            cur_a: Vec::new(),
            started: false,
        }
    }
    fn flush(&mut self) {
        if self.started {
            self.fields.push((
                core::mem::take(&mut self.cur_c),
                core::mem::take(&mut self.cur_a),
            ));
            self.started = false;
        }
    }
    /// Non-splittable text (literal or quoted expansion).
    fn push_unsplit(&mut self, s: &str, glob_active: bool) {
        for ch in s.chars() {
            self.cur_c.push(ch);
            self.cur_a.push(glob_active);
        }
        self.started = true;
    }
    /// Splittable text (unquoted expansion) — break on IFS.
    fn push_split(&mut self, s: &str, glob_active: bool, ifs: &[char]) {
        let mut i = 0;
        let cs: Vec<char> = s.chars().collect();
        while i < cs.len() {
            if ifs.contains(&cs[i]) {
                if self.started {
                    self.flush();
                }
                while i < cs.len() && ifs.contains(&cs[i]) {
                    i += 1;
                }
            } else {
                self.cur_c.push(cs[i]);
                self.cur_a.push(glob_active);
                self.started = true;
                i += 1;
            }
        }
    }
    fn finish(mut self) -> Vec<(Vec<char>, Vec<bool>)> {
        self.flush();
        self.fields
    }
}

/// Expand a word into zero or more fields, with field splitting and pathname
/// expansion applied. Used for command arguments.
pub fn expand_to_fields(word: &Word, ctx: &mut dyn ExpandCtx) -> R<Vec<String>> {
    let ifs_str = ctx.ifs();
    let ifs: Vec<char> = ifs_str.chars().collect();
    let mut b = FieldBuilder::new();
    let word = apply_tilde(word, ctx);

    for part in &word {
        match part {
            WordPart::Lit { text, from_quote } => b.push_unsplit(text, !from_quote),
            WordPart::Sub { raw, quoted } => {
                let out = ctx.command_subst(raw);
                if *quoted {
                    b.push_unsplit(&out, false);
                } else {
                    b.push_split(&out, true, &ifs);
                }
            }
            WordPart::Arith { raw, quoted } => {
                let v = ctx.arith(raw).to_string();
                if *quoted {
                    b.push_unsplit(&v, false);
                } else {
                    b.push_split(&v, true, &ifs);
                }
            }
            WordPart::Var { name, op, quoted } => {
                expand_var_into(&mut b, name, op, *quoted, ctx, &ifs)?;
            }
        }
    }

    let raw_fields = b.finish();
    // Pathname expansion per field.
    let cwd = ctx.cwd();
    let mut out: Vec<String> = Vec::new();
    for (chars, active) in raw_fields {
        let mut lister = |d: &str| ctx.list_dir(d);
        let matches = expand_glob_masked(&chars, &active, &cwd, &mut lister);
        out.extend(matches);
    }
    Ok(out)
}

/// Expand a word to a single string (no field splitting, no globbing). Used for
/// assignment values, redirect targets, `case` subjects, and `${…}` op words.
pub fn expand_to_string(word: &Word, ctx: &mut dyn ExpandCtx) -> R<String> {
    let word = apply_tilde(word, ctx);
    let mut s = String::new();
    for part in &word {
        match part {
            WordPart::Lit { text, .. } => s.push_str(text),
            WordPart::Sub { raw, .. } => s.push_str(&ctx.command_subst(raw)),
            WordPart::Arith { raw, .. } => s.push_str(&ctx.arith(raw).to_string()),
            WordPart::Var { name, op, .. } => s.push_str(&scalar_var(name, op, ctx)?),
        }
    }
    Ok(s)
}

/// Expand a redirect target to exactly one field (POSIX "ambiguous redirect").
/// Errors are flattened to a `String` so the redirection layer can report them
/// uniformly with its own file-open errors.
pub fn expand_redirect_target(word: &Word, ctx: &mut dyn ExpandCtx) -> Result<String, String> {
    let fields = expand_to_fields(word, ctx).map_err(|e| e.0)?;
    match fields.len() {
        1 => Ok(fields.into_iter().next().unwrap()),
        0 => Err("ambiguous redirect: empty".to_string()),
        _ => Err("ambiguous redirect: multiple files".to_string()),
    }
}

fn expand_var_into(
    b: &mut FieldBuilder,
    name: &str,
    op: &ParamOp,
    quoted: bool,
    ctx: &mut dyn ExpandCtx,
    ifs: &[char],
) -> R<()> {
    // $@ / $* — multi-field special handling.
    if matches!(name, "@" | "*") && matches!(op, ParamOp::Get) {
        let params = ctx.positionals();
        if name == "*" && quoted {
            // "$*" → single field, joined by first IFS char (space if empty).
            let sep = ifs.first().copied().unwrap_or(' ');
            let joined = params.join(&sep.to_string());
            b.push_unsplit(&joined, false);
            return Ok(());
        }
        // "$@" (quoted) → one field per param; $@/$* unquoted → split each.
        for (k, p) in params.iter().enumerate() {
            if k > 0 {
                b.flush();
            }
            if quoted {
                b.push_unsplit(p, false);
            } else {
                b.push_split(p, true, ifs);
            }
        }
        return Ok(());
    }
    // ${#@} / ${#*} → positional count.
    if matches!(op, ParamOp::Length) && matches!(name, "@" | "*") {
        let n = ctx.positionals().len();
        let s = n.to_string();
        if quoted {
            b.push_unsplit(&s, false);
        } else {
            b.push_split(&s, true, ifs);
        }
        return Ok(());
    }
    let val = scalar_var(name, op, ctx)?;
    if quoted {
        b.push_unsplit(&val, false);
    } else {
        b.push_split(&val, true, ifs);
    }
    Ok(())
}

/// Compute the string value of a scalar parameter expansion.
fn scalar_var(name: &str, op: &ParamOp, ctx: &mut dyn ExpandCtx) -> R<String> {
    let base = lookup(ctx, name);
    Ok(match op {
        ParamOp::Get => base.unwrap_or_default(),
        ParamOp::Length => base.unwrap_or_default().chars().count().to_string(),
        ParamOp::Default { colon, word } => {
            if is_unset_or_null(&base, *colon) {
                expand_to_string(word, ctx)?
            } else {
                base.unwrap_or_default()
            }
        }
        ParamOp::Assign { colon, word } => {
            if is_unset_or_null(&base, *colon) {
                let v = expand_to_string(word, ctx)?;
                if assignable(name) {
                    ctx.set(name, &v);
                }
                v
            } else {
                base.unwrap_or_default()
            }
        }
        ParamOp::Alt { colon, word } => {
            if is_unset_or_null(&base, *colon) {
                String::new()
            } else {
                expand_to_string(word, ctx)?
            }
        }
        ParamOp::Error { colon, word } => {
            if is_unset_or_null(&base, *colon) {
                let msg = expand_to_string(word, ctx)?;
                return Err(ExpandError(if msg.is_empty() {
                    alloc::format!("{name}: parameter null or not set")
                } else {
                    alloc::format!("{name}: {msg}")
                }));
            } else {
                base.unwrap_or_default()
            }
        }
        ParamOp::TrimPrefix { longest, pat } => {
            let p = expand_to_string(pat, ctx)?;
            trim_prefix(&base.unwrap_or_default(), &p, *longest)
        }
        ParamOp::TrimSuffix { longest, pat } => {
            let p = expand_to_string(pat, ctx)?;
            trim_suffix(&base.unwrap_or_default(), &p, *longest)
        }
    })
}

fn is_unset_or_null(base: &Option<String>, colon: bool) -> bool {
    match base {
        None => true,
        Some(v) => colon && v.is_empty(),
    }
}

fn assignable(name: &str) -> bool {
    !name.is_empty()
        && !name.starts_with(|c: char| c.is_ascii_digit())
        && name.chars().all(|c| c == '_' || c.is_ascii_alphanumeric())
}

/// Leading unquoted `~` / `~/...` → $HOME. (`~user` left literal: no passwd db.)
fn apply_tilde(word: &Word, ctx: &mut dyn ExpandCtx) -> Word {
    if let Some(WordPart::Lit {
        text,
        from_quote: false,
    }) = word.first()
    {
        if text.starts_with('~') && (text.len() == 1 || text.as_bytes()[1] == b'/') {
            if let Some(home) = ctx.home() {
                let mut new = word.clone();
                let rest = &text[1..];
                new[0] = WordPart::Lit {
                    text: alloc::format!("{home}{rest}"),
                    from_quote: false,
                };
                return new;
            }
        }
    }
    word.clone()
}

// ---------- pattern trimming (uses glob matcher) ----------

fn trim_prefix(value: &str, pat: &str, longest: bool) -> String {
    if pat.is_empty() {
        return value.to_string();
    }
    let s: Vec<char> = value.chars().collect();
    let p: Vec<char> = pat.chars().collect();
    let mut best: Option<usize> = None;
    for i in 0..=s.len() {
        if crate::glob::glob_full(&p, &s[..i]) {
            best = Some(i);
            if !longest {
                break;
            }
        }
    }
    match best {
        Some(i) => s[i..].iter().collect(),
        None => value.to_string(),
    }
}

fn trim_suffix(value: &str, pat: &str, longest: bool) -> String {
    if pat.is_empty() {
        return value.to_string();
    }
    let s: Vec<char> = value.chars().collect();
    let p: Vec<char> = pat.chars().collect();
    let mut best: Option<usize> = None;
    for j in 0..=s.len() {
        if crate::glob::glob_full(&p, &s[s.len() - j..]) {
            best = Some(j);
            if !longest {
                break;
            }
        }
    }
    match best {
        Some(j) => s[..s.len() - j].iter().collect(),
        None => value.to_string(),
    }
}

File: memcontainers/shcore/src/expand.zig
//! Word expansion: tilde, parameter/command/arithmetic, IFS splitting, globbing.

const std = @import("std");
const glob = @import("glob.zig");
const word = @import("word.zig");

pub const ExpandError = std.mem.Allocator.Error || error{
    AmbiguousRedirectEmpty,
    AmbiguousRedirectMultiple,
    Parameter,
};

pub const ExpandContext = struct {
    ptr: *anyopaque,
    vtable: *const VTable,

    pub const VTable = struct {
        get: *const fn (*anyopaque, std.mem.Allocator, []const u8) ?[]const u8,
        set: *const fn (*anyopaque, []const u8, []const u8) void,
        special: *const fn (*anyopaque, std.mem.Allocator, []const u8) ?[]const u8,
        positionals: *const fn (*anyopaque, std.mem.Allocator) []const []const u8,
        command_subst: *const fn (*anyopaque, std.mem.Allocator, []const u8) []const u8,
        arith: *const fn (*anyopaque, std.mem.Allocator, []const u8) i64,
        list_dir: *const fn (*anyopaque, std.mem.Allocator, []const u8) ?[]const []const u8,
        cwd: *const fn (*anyopaque, std.mem.Allocator) []const u8,
        ifs: *const fn (*anyopaque, std.mem.Allocator) []const u8,
        home: *const fn (*anyopaque, std.mem.Allocator) ?[]const u8,
    };

    pub fn get(self: *ExpandContext, allocator: std.mem.Allocator, name: []const u8) ?[]const u8 {
        if (self.vtable.special(self.ptr, allocator, name)) |v| return v;
        return self.vtable.get(self.ptr, allocator, name);
    }

    pub fn set(self: *ExpandContext, name: []const u8, value: []const u8) void {
        self.vtable.set(self.ptr, name, value);
    }

    pub fn positionals(self: *ExpandContext, allocator: std.mem.Allocator) []const []const u8 {
        return self.vtable.positionals(self.ptr, allocator);
    }

    pub fn commandSubst(self: *ExpandContext, allocator: std.mem.Allocator, raw: []const u8) []const u8 {
        return self.vtable.command_subst(self.ptr, allocator, raw);
    }

    pub fn arith(self: *ExpandContext, allocator: std.mem.Allocator, raw: []const u8) i64 {
        return self.vtable.arith(self.ptr, allocator, raw);
    }

    pub fn listDir(self: *ExpandContext, allocator: std.mem.Allocator, path: []const u8) ?[]const []const u8 {
        return self.vtable.list_dir(self.ptr, allocator, path);
    }

    pub fn cwd(self: *ExpandContext, allocator: std.mem.Allocator) []const u8 {
        return self.vtable.cwd(self.ptr, allocator);
    }

    pub fn ifs(self: *ExpandContext, allocator: std.mem.Allocator) []const u8 {
        return self.vtable.ifs(self.ptr, allocator);
    }

    pub fn home(self: *ExpandContext, allocator: std.mem.Allocator) ?[]const u8 {
        return self.vtable.home(self.ptr, allocator);
    }
};

const Field = struct {
    text: []const u8,
    active: []const bool,
};

const FieldBuilder = struct {
    fields: std.ArrayList(Field) = .empty,
    cur_text: std.ArrayList(u8) = .empty,
    cur_active: std.ArrayList(bool) = .empty,
    started: bool = false,

    fn flush(self: *FieldBuilder, allocator: std.mem.Allocator) !void {
        if (!self.started) return;
        try self.fields.append(allocator, .{
            .text = try self.cur_text.toOwnedSlice(allocator),
            .active = try self.cur_active.toOwnedSlice(allocator),
        });
        self.cur_text = .empty;
        self.cur_active = .empty;
        self.started = false;
    }

    fn pushUnsplit(self: *FieldBuilder, allocator: std.mem.Allocator, bytes: []const u8, glob_active: bool) !void {
        try self.cur_text.appendSlice(allocator, bytes);
        for (bytes) |_| try self.cur_active.append(allocator, glob_active);
        self.started = true;
    }

    fn pushSplit(self: *FieldBuilder, allocator: std.mem.Allocator, bytes: []const u8, glob_active: bool, ifs: []const u8) !void {
        var i: usize = 0;
        while (i < bytes.len) {
            if (containsByte(ifs, bytes[i])) {
                if (self.started) try self.flush(allocator);
                while (i < bytes.len and containsByte(ifs, bytes[i])) i += 1;
                continue;
            }
            try self.cur_text.append(allocator, bytes[i]);
            try self.cur_active.append(allocator, glob_active);
            self.started = true;
            i += 1;
        }
    }

    fn finish(self: *FieldBuilder, allocator: std.mem.Allocator) ![]const Field {
        try self.flush(allocator);
        return self.fields.toOwnedSlice(allocator);
    }
};

pub fn expandToFields(allocator: std.mem.Allocator, input: word.Word, ctx: *ExpandContext) ExpandError![]const []const u8 {
    var arena = std.heap.ArenaAllocator.init(allocator);
    defer arena.deinit();
    const scratch = arena.allocator();

    const ifs = ctx.ifs(scratch);
    const expanded_word = try applyTilde(scratch, input, ctx);
    var builder = FieldBuilder{};

    for (expanded_word) |part| {
        switch (part) {
            .lit => |lit| try builder.pushUnsplit(scratch, lit.text, !lit.from_quote),
            .sub => |sub| {
                const out = ctx.commandSubst(scratch, sub.raw);
                if (sub.quoted) {
                    try builder.pushUnsplit(scratch, out, false);
                } else {
                    try builder.pushSplit(scratch, out, true, ifs);
                }
            },
            .arith => |arith_part| {
                const value = try std.fmt.allocPrint(scratch, "{d}", .{ctx.arith(scratch, arith_part.raw)});
                if (arith_part.quoted) {
                    try builder.pushUnsplit(scratch, value, false);
                } else {
                    try builder.pushSplit(scratch, value, true, ifs);
                }
            },
            .param => |param| try expandVarInto(scratch, &builder, param.name, param.op, param.quoted, ctx, ifs),
        }
    }

    const raw_fields = try builder.finish(scratch);
    const cwd = ctx.cwd(scratch);
    var out = std.ArrayList([]const u8).empty;
    errdefer freePendingStrings(allocator, &out);
    for (raw_fields) |field| {
        const matches = try glob.expandGlobMasked(
            scratch,
            field.text,
            field.active,
            cwd,
            @ptrCast(ctx),
            listDirAdapter,
        );
        for (matches) |match| {
            const owned = try allocator.dupe(u8, match);
            out.append(allocator, owned) catch |err| {
                allocator.free(owned);
                return err;
            };
        }
    }
    return out.toOwnedSlice(allocator);
}

pub fn expandToString(allocator: std.mem.Allocator, input: word.Word, ctx: *ExpandContext) ExpandError![]const u8 {
    var arena = std.heap.ArenaAllocator.init(allocator);
    defer arena.deinit();
    const scratch = arena.allocator();

    const expanded_word = try applyTilde(scratch, input, ctx);
    var out = std.ArrayList(u8).empty;
    for (expanded_word) |part| {
        switch (part) {
            .lit => |lit| try out.appendSlice(scratch, lit.text),
            .sub => |sub| try out.appendSlice(scratch, ctx.commandSubst(scratch, sub.raw)),
            .arith => |arith_part| {
                const value = try std.fmt.allocPrint(scratch, "{d}", .{ctx.arith(scratch, arith_part.raw)});
                try out.appendSlice(scratch, value);
            },
            .param => |param| try out.appendSlice(scratch, try scalarVar(scratch, param.name, param.op, ctx)),
        }
    }
    return allocator.dupe(u8, out.items);
}

pub fn expandRedirectTarget(allocator: std.mem.Allocator, input: word.Word, ctx: *ExpandContext) ExpandError![]const u8 {
    const fields = try expandToFields(allocator, input, ctx);
    return switch (fields.len) {
        0 => blk: {
            if (fields.len != 0) allocator.free(fields);
            break :blk error.AmbiguousRedirectEmpty;
        },
        1 => blk: {
            const out = fields[0];
            allocator.free(fields);
            break :blk out;
        },
        else => blk: {
            freeStringList(allocator, fields);
            break :blk error.AmbiguousRedirectMultiple;
        },
    };
}

fn freePendingStrings(allocator: std.mem.Allocator, list: *std.ArrayList([]const u8)) void {
    for (list.items) |item| allocator.free(item);
    list.deinit(allocator);
}

fn freeStringList(allocator: std.mem.Allocator, items: []const []const u8) void {
    for (items) |item| allocator.free(item);
    allocator.free(items);
}

fn expandVarInto(
    allocator: std.mem.Allocator,
    builder: *FieldBuilder,
    name: []const u8,
    op: word.ParamOp,
    quoted: bool,
    ctx: *ExpandContext,
    ifs: []const u8,
) ExpandError!void {
    if ((std.mem.eql(u8, name, "@") or std.mem.eql(u8, name, "*")) and op == .get) {
        const params = ctx.positionals(allocator);
        if (std.mem.eql(u8, name, "*") and quoted) {
            try builder.pushUnsplit(allocator, try joinPositionals(allocator, params, ifs), false);
            return;
        }
        for (params, 0..) |param, idx| {
            if (idx > 0) try builder.flush(allocator);
            if (quoted) {
                try builder.pushUnsplit(allocator, param, false);
            } else {
                try builder.pushSplit(allocator, param, true, ifs);
            }
        }
        return;
    }

    if ((std.mem.eql(u8, name, "@") or std.mem.eql(u8, name, "*")) and op == .length) {
        const len = ctx.positionals(allocator).len;
        const value = try std.fmt.allocPrint(allocator, "{d}", .{len});
        if (quoted) {
            try builder.pushUnsplit(allocator, value, false);
        } else {
            try builder.pushSplit(allocator, value, true, ifs);
        }
        return;
    }

    const value = try scalarVar(allocator, name, op, ctx);
    if (quoted) {
        try builder.pushUnsplit(allocator, value, false);
    } else {
        try builder.pushSplit(allocator, value, true, ifs);
    }
}

fn scalarVar(allocator: std.mem.Allocator, name: []const u8, op: word.ParamOp, ctx: *ExpandContext) ExpandError![]const u8 {
    const base = ctx.get(allocator, name);
    return switch (op) {
        .get => if (base) |v| v else "",
        .length => blk: {
            const v = base orelse "";
            break :blk try std.fmt.allocPrint(allocator, "{d}", .{v.len});
        },
        .default_value => |param| if (isUnsetOrNull(base, param.colon))
            try expandToString(allocator, param.word, ctx)
        else
            base.?,
        .assign => |param| blk: {
            if (isUnsetOrNull(base, param.colon)) {
                const value = try expandToString(allocator, param.word, ctx);
                if (assignable(name)) ctx.set(name, value);
                break :blk value;
            }
            break :blk base.?;
        },
        .alt => |param| if (isUnsetOrNull(base, param.colon))
            ""
        else
            try expandToString(allocator, param.word, ctx),
        .err => |param| {
            if (isUnsetOrNull(base, param.colon)) return error.Parameter;
            return base.?;
        },
        .trim_prefix => |trim| blk: {
            const pat = try expandToString(allocator, trim.pat, ctx);
            break :blk try trimPrefix(allocator, base orelse "", pat, trim.longest);
        },
        .trim_suffix => |trim| blk: {
            const pat = try expandToString(allocator, trim.pat, ctx);
            break :blk try trimSuffix(allocator, base orelse "", pat, trim.longest);
        },
    };
}

fn applyTilde(allocator: std.mem.Allocator, input: word.Word, ctx: *ExpandContext) ExpandError!word.Word {
    if (input.len == 0) return input;
    const lit = switch (input[0]) {
        .lit => |lit| lit,
        else => return input,
    };
    if (lit.from_quote or lit.text.len == 0 or lit.text[0] != '~') return input;
    if (!(lit.text.len == 1 or lit.text[1] == '/')) return input;
    const home = ctx.home(allocator) orelse return input;

    const out = try allocator.alloc(word.WordPart, input.len);
    @memcpy(out, input);
    out[0] = .{ .lit = .{
        .text = try std.fmt.allocPrint(allocator, "{s}{s}", .{ home, lit.text[1..] }),
        .from_quote = false,
    } };
    return out;
}

fn trimPrefix(allocator: std.mem.Allocator, value: []const u8, pat: []const u8, longest: bool) ![]const u8 {
    if (pat.len == 0) return allocator.dupe(u8, value);
    var best: ?usize = null;
    var i: usize = 0;
    while (i <= value.len) : (i += 1) {
        if (glob.globFull(pat, value[0..i])) {
            best = i;
            if (!longest) break;
        }
    }
    if (best) |idx| return allocator.dupe(u8, value[idx..]);
    return allocator.dupe(u8, value);
}

fn trimSuffix(allocator: std.mem.Allocator, value: []const u8, pat: []const u8, longest: bool) ![]const u8 {
    if (pat.len == 0) return allocator.dupe(u8, value);
    var best: ?usize = null;
    var j: usize = 0;
    while (j <= value.len) : (j += 1) {
        if (glob.globFull(pat, value[value.len - j ..])) {
            best = j;
            if (!longest) break;
        }
    }
    if (best) |len| return allocator.dupe(u8, value[0 .. value.len - len]);
    return allocator.dupe(u8, value);
}

fn joinPositionals(allocator: std.mem.Allocator, params: []const []const u8, ifs: []const u8) ![]const u8 {
    const sep: []const u8 = if (ifs.len == 0) " " else ifs[0..1];
    var out = std.ArrayList(u8).empty;
    for (params, 0..) |param, idx| {
        if (idx > 0) try out.appendSlice(allocator, sep);
        try out.appendSlice(allocator, param);
    }
    return out.toOwnedSlice(allocator);
}

fn isUnsetOrNull(base: ?[]const u8, colon: bool) bool {
    const v = base orelse return true;
    return colon and v.len == 0;
}

fn assignable(name: []const u8) bool {
    if (name.len == 0) return false;
    if (std.ascii.isDigit(name[0])) return false;
    for (name) |ch| {
        if (!(ch == '_' or std.ascii.isAlphanumeric(ch))) return false;
    }
    return true;
}

fn containsByte(haystack: []const u8, needle: u8) bool {
    for (haystack) |ch| {
        if (ch == needle) return true;
    }
    return false;
}

fn listDirAdapter(ptr: *anyopaque, allocator: std.mem.Allocator, path: []const u8) ?[]const []const u8 {
    const ctx: *ExpandContext = @ptrCast(@alignCast(ptr));
    return ctx.listDir(allocator, path);
}
